DSDEGP-2009: Create key, metatdata, query-input, and query-output classes for Arrays

tbl3rd opened the pull request:
The TransferIndex classes for Arrays compile now.
I'm sure they are wrong, but don't know how to make them right.

### Description

_Please explain the changes you made here._
_Explain the **motivation** for making this change. What existing problem does the pull request solve?_
_Mention any JIRA issues fixed, addressed or otherwise related to this pull request, including issue numbers or hard links for issues in other repos._

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: Ensure that you have added or updated tests
- [ ] **Submitter**: If you're adding/switching libraries or otherwise changing the design, update the [design documentation](https://broadinstitute.atlassian.net/wiki/pages/viewpage.action?pageId=114531509).
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
* Review cycle:
  * Team may comment on PR at will
  * **Reviewer assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to reviewer** for further feedback
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Move the JIRA issue to QA once this checklist is completed

tbl3rd committed:
The Arrays transfer index classes compile.

tbl3rd committed:
The idats files file use the .idats.txt extension.

tbl3rd committed:
Oh, thank you, Scalafmt!

danxmoran commented on clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/arrays/TransferArraysV1Metadata.scala:
```diff
@@ -0,0 +1,111 @@
+  autocallGender: Option[Symbol] = None,
+  autocallPf: Option[Boolean] = None,
+  autocallVersion: Option[Int] = None,
+  beadPoolManifestFile: Option[URI] = None,
```
In other indexes we usually suffix these with `Path` instead of `File`

danxmoran commented on clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/arrays/TransferArraysV1Metadata.scala:
```diff
@@ -0,0 +1,111 @@
+  autocallVersion: Option[Int] = None,
+  beadPoolManifestFile: Option[URI] = None,
+  callRate: Option[Double] = None,
+  callRateThreshold: Option[Float] = None,
```
Curious: why `Float` for this one, instead of `Double`?

gbggrant replied:
Don't need to store the callRateThreshold.  It's used by the workflow to make a QC decision on the workflow.

jacarey replied:
I'm not sure we need the precision of a Double on any of these metrics to TBH

danxmoran commented on clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/arrays/TransferArraysV1Metadata.scala:
```diff
@@ -0,0 +1,111 @@
+      refFasta,
+      refFastaIndex,
+      zcallThresholdsFile
+    )
```
Do we want all of these to get wiped out by the delete CLP, or do we want to leave some around (like we do with metrics files for the cram / gvcf indexes)?

tbl3rd replied:
I don't know. Probably. Does anyone know the answer now? Or can it wait?

danxmoran replied:
@ktibbett might know... but it can also wait, as long as we remember to fix it up before we release the CLPs.

danxmoran commented on clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/arrays/TransferArraysV1Metadata.scala:
```diff
@@ -0,0 +1,111 @@
+    refFasta = pathMapper(refFasta, ArraysExtensions.FastaExtension),
+    refFastaIndex = pathMapper(refFastaIndex, ArraysExtensions.FastaFaiExtension),
+    zcallThresholdsFile =
+      pathMapper(zcallThresholdsFile, ArraysExtensions.EgtThresholdsTxtExtension)
```
An extra twist: this `copy` is used to build the set of moves that should happen for data-delivery. Not sure if there's anything we should leave out, like we do with cram metrics.

danxmoran replied:
@ktibbett probably also knows the list of things that should be delivered and what shouldn't move around. From working on the delivery side of things, it looks like we'll want to deliver:
1. vcf
2. idat(s)
3. gtc

And any indexes for those things (I'm guessing).

tbl3rd committed:
Make it easier to review via deletion rather than addition.

tbl3rd committed:
gtc -=> gtcFile

gbggrant commented on clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/arrays/ArraysExtensions.scala:
```diff
@@ -0,0 +1,41 @@
+
+  val GzExtension = ".gz"
+
+  val IdatsExtension = ".idats"
```
should be idat extension.  There are two idat files that we need to store here.  The workflow uses the idats file, but it's contents are the paths to the red and green idats.  We don't need to store anything about the idats file if you get what I mean.

tbl3rd replied:
Not sure I get what you mean about which files to store. I was purposely redundant here to make it easier to just delete what we don't want rather than have to discover what we do want. We obviously don't have to store both the file containing the idat paths and the idat paths themselves. I think it would be simpler to just store the two .idat files, but want confirmation.

gbggrant commented on clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/arrays/ArraysExtensions.scala:
```diff
@@ -0,0 +1,41 @@
+
+  val CsvExtension = ".csv"
+
+  val DictExtension = ".dict"
```
Many of these extensions are not arrays specific.

pshapiro4broad replied:
I would just make this `FileExtension` and not try to organize extensions by index type. Using singular rather than plural and dropping the redundant `Extension` from the name would make them easier to read, IMO, e.g. `FileExtension.Bpm` instead of `FileExtensions.BpmExtension`.

gbggrant commented on clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/arrays/ArraysExtensions.scala:
```diff
@@ -0,0 +1,41 @@
+
+  val EgtThresholdsTxtExtension = s"${EgtExtension}${ThresholdsExtension}${TxtExtension}"
+
+  val IdatsTxtExtension = s"${IdatsExtension}${TxtExtension}"
```
idats -> idat

gbggrant commented on clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/arrays/TransferArraysV1Metadata.scala:
```diff
@@ -0,0 +1,151 @@
+
+case class TransferArraysV1Metadata(
+  autocallCallRate: Option[Double] = None,
+  autocallDate: Option[OffsetDateTime] = None,
```
why OffsetDateTime?

tbl3rd replied:
Why not OffsetDateTime?

pshapiro4broad replied:
`OffsetDateTime` is what we currently use to store dates in elasticsearch.

gbggrant commented on clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/arrays/TransferArraysV1Metadata.scala:
```diff
@@ -0,0 +1,151 @@
+  chipType: Option[String] = None,
+  clusterFile: Option[URI] = None,
+  controlSampleName: Option[String] = None,
+  createdAt: Option[OffsetDateTime] = None,
```
These are timestamps in the database for when the row was created.  Don't we have this in clio

pshapiro4broad replied:
Is the idea here to preserve the creation time from arrays database row? Not sure what the use of this field is. Anyway, elasticsearch itself doesn't automatically create a 'created at' property for you, and I don't think clio has this concept either.

tbl3rd replied:
Aside from mistakes: the fields all come from some Arrays data source I consulted. CREATED_AT is a column in the ARRAYS_QC table in the cloud metrics database. (I'll drop the fields derived from the *_AT entries unless someone objects.)

gbggrant commented on clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/arrays/TransferArraysV1Metadata.scala:
```diff
@@ -0,0 +1,151 @@
+  imagingDate: Option[OffsetDateTime] = None,
+  isLatest: Option[Boolean] = None,
+  isZcalled: Option[Boolean] = None,
+  modifiedAt: Option[OffsetDateTime] = None,
```
These are timestamps in the database for when the row was created.  Don't we have this in clio?

gbggrant commented on clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/arrays/TransferArraysV1QueryInput.scala:
```diff
@@ -0,0 +1,30 @@
+  pipelineVersion: Option[Symbol] = None,
+  workflowStartDate: Option[OffsetDateTime] = None,
+  workflowEndDate: Option[OffsetDateTime] = None,
+  cramMd5: Option[Symbol] = None,
```
No Crams in arrays

gbggrant commented on clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/arrays/TransferArraysV1QueryInput.scala:
```diff
@@ -0,0 +1,30 @@
+  cromwellId: Option[UUID] = None,
+  readgroupMd5: Option[Symbol] = None,
+  workspaceName: Option[String] = None,
+  regulatoryDesignation: Option[RegulatoryDesignation] = None,
```
I don't think we use regulatoryDesignation in arrays

pshapiro4broad replied:
We don't think we'll ever do clinical arrays? Or are you saying that we don't have that functionality now, so why add it now?

jacarey commented on clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/arrays/TransferArraysV1Key.scala:
```diff
@@ -0,0 +1,20 @@
+  override def getUrlSegments: Seq[String] =
+    Seq(
+      location.entryName,
+      chipwellBarcode,
```
We may want to separate chip barcode from chip position. It could help with chip based queries. Thoughts?

tbl3rd replied:
Please explain more, Jay.

jacarey replied:
chip well barcode is actually and combination of the chip barcode and the well position.  so 123456789_R01C02 is chip barcode 123456789 row 01 column 02

If someone noticed something was wrong with a sample they might want to go back and look at other samples on that same chip. So we could either keep chipwell barcode as is and look for all keys that are like 123456789_* or we could separate the chip barcode from the well position and just search for all samples with a chip barcode of 123456789.

jacarey commented on clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/arrays/TransferArraysV1QueryInput.scala:
```diff
@@ -0,0 +1,30 @@
+  cramSize: Option[Long] = None,
+  cramPath: Option[URI] = None,
+  cromwellId: Option[UUID] = None,
+  readgroupMd5: Option[Symbol] = None,
```
no readgroups either.

tbl3rd committed:
Respond to nearly all review comments.

Still gotta figure out what Jay meant about "chip position"
and what CopyArraysToBucket does with MultiSampleArrays data.

tbl3rd committed:
Type chipwellBarcode as Symbol instead of String.

Fewer surprises matching a regular expression to chipwell location.
